crates/protocol/src/payload.rs

```diff
@@ -0,0 +1,314 @@
+
+/// Memory overhead per payload transaction.
+/// 24 bytes per tx overhead (size of slice header in memory).
+pub const PAYLOAD_TX_MEM_OVERHEAD: u64 = 24;
+use super::{BlockInfo, L1BlockInfoBedrock, L1BlockInfoEcotone, L1BlockInfoTx, L2BlockInfo};
+use crate::block::OpBlock;
+
+use superchain_primitives::{RollupConfig, SystemConfig};
+
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+/// The error for execution payload
+#[derive(Debug)]
+pub enum ExecutionPayloadError {
+    /// invalid genesis hash in the execution payload.
+    InvalidGenesisHash,
+    /// Missing L1 deposit transaction in an L2 block.
+    MissingL1InfoDeposit(B256),
+    ///  Unexpected transaction type was encountered.
+    UnexpectedTransactionType(String),
+    /// Decoding process error
+    DecodingError(String),
+    /// missing system configuration in the genesis block.
+    MissingSystemConfig,
+}
+impl core::fmt::Display for ExecutionPayloadError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            ExecutionPayloadError::InvalidGenesisHash => {
+                write!(f, "Invalid genesis hash for the payload")
+            }
+            ExecutionPayloadError::MissingL1InfoDeposit(hash) => {
+                write!(f, "Missing L1 information deposit transaction for block hash: {:?}", hash)
+            }
+            ExecutionPayloadError::UnexpectedTransactionType(ty) => {
+                write!(f, "Unexpected transaction type: {:?}", ty)
+            }
+            ExecutionPayloadError::DecodingError(e) => write!(f, "Error decoding data: {}", e),
+            ExecutionPayloadError::MissingSystemConfig => {
+                write!(f, "System configuration missing in genesis block")
+            }
+        }
+    }
+}
+#[cfg(not(feature = "std"))]
+impl std::error::Error for ExecutionPayloadError {}
+
+type Result<T> = core::result::Result<T, ExecutionPayloadError>;
+/// Envelope wrapping the [L2ExecutionPayload].
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct L2ExecutionPayloadEnvelope {
+    /// Parent beacon block root.
+    #[cfg_attr(feature = "serde", serde(rename = "parentBeaconBlockRoot"))]
+    pub parent_beacon_block_root: Option<B256>,
+    /// The inner execution payload.
+    #[cfg_attr(feature = "serde", serde(rename = "executionPayload"))]
+    pub execution_payload: L2ExecutionPayload,
+}
+
+impl L2ExecutionPayloadEnvelope {
+    /// Returns the payload memory size.
+    pub fn mem_size(&self) -> u64 {
+        let mut out = PAYLOAD_MEM_FIXED_COST;
+        for tx in &self.execution_payload.transactions {
+            out += tx.len() as u64 + PAYLOAD_TX_MEM_OVERHEAD;
+        }
+        out
+    }
+}
+
+/// The execution payload.
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct L2ExecutionPayload {
```

Comment:
We should be able to avoid duplicating this type and use an alloy execution payload type as well.

Reply:
Here's the `ExecutionPayload` type: https://github.com/alloy-rs/alloy/blob/4492c25f1a666881d18367dc47d0040fc05b17ae/crates/rpc-types-engine/src/payload.rs#L594

Reply:
I also have a pr open in upstream alloy to support `no_std` in `alloy-rpc-types-engine`: https://github.com/alloy-rs/alloy/pull/1233

Reply:
Thanks for sharing this,

 Should we wait for your changes to get merged upstream before we integrate this into the crate?